File: Sorting-Algorithms/2-Merge Sort/Merge Sort.py
#--------------------------------------------------------------------------
#                       MERGE SORT
#--------------------------------------------------------------------------

def merge(arr, left, mid, right):
    n1 = mid - left + 1
    n2 = right - mid

    left_arr = [0] * n1
    right_arr = [0] * n2

    for i in range(n1):
        left_arr[i] = arr[left + i]
    for j in range(n2):
        right_arr[j] = arr[mid + 1 + j]

    i = 0       # In leftArr
    j = 0       # In rightArr
    k = left    # Main Array

    # Merge two temporary arrays
    while i < n1 and j < n2:
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1

    # Copy the remaining elements of leftArr (if any)
    while i < n1:
        arr[k] = left_arr[i]
        i += 1
        k += 1
    # Copy the remaining elements of rightArr (if any)
    while j < n2:
        arr[k] = right_arr[j]
        j += 1
        k += 1


def merge_sort(arr, left, right):
    if left >= right:
        return

    mid = (left + right) // 2

    # Recursive division
    merge_sort(arr, left, mid)
    merge_sort(arr, mid + 1, right)

    # Merging Two Half Sorted
    merge(arr, left, mid, right)



if __name__ == "__main__":
    arr = [64, 45, 48, 87, 878, 78, 74, 13, 55, 44, 34, 25, 12, 22, 100, 210, 150, 124, 11, 90]
    print("BEFORE SORTING :", arr)

    merge_sort(arr, 0, len(arr) - 1)

    print("AFTER SORTED  :", arr)

File: Sorting-Algorithms/2-Merge Sort/Merge Sort.cpp
//---------------------------------------------------------------------
//                      MERGE SORT
//---------------------------------------------------------------------

#include <iostream>
#include <vector>
using namespace std;

void merge(vector<int>& arr, int left, int mid, int right)
{
    int n1 = mid - left + 1;    // Left Half
    int n2 = right - mid;       // Right Half


    vector<int> leftArr(n1), rightArr(n2);

    for(int i = 0; i < n1; i++)
        leftArr[i] = arr[left + i];
    for(int j = 0; j < n2; j++)
        rightArr[j] = arr[mid + 1 + j];


    // Start 3 Andis for all 3 Arrays
    int i = 0;          // In leftArr
    int j = 0;          // In rightArr
    int k = left;       // Main Array


    // Merge two temporary arrays in ascending order
    while (i < n1 && j < n2)
    {
        if(leftArr[i] <= rightArr[j])
        {
            arr[k] = leftArr[i];
            i++;
        }

        else
        {
            arr[k] = rightArr[j];
            j++;
        }
        k++;
    }

    // Copy the remaining elements of leftArr (if any)
    while (i < n1)
    {
        arr[k] = leftArr[i];
        i++;
        k++;
    }

    // Copy the remaining elements of rightArr (if any)
    while (j < n2)
    {
        arr[k] = rightArr[j];
        j++;
        k++;
    }
}


// Main Merge Sort Func
void mergeSort(vector<int>& arr, int left, int right)
{
    if(left >= right)
        return;

    int mid = left + (right - left) / 2;    // Protected From Buffering

    // Recursive division
    mergeSort(arr, left, mid);
    mergeSort(arr, mid + 1, right);

    // Merging Two Half Sorted
    merge(arr, left, mid, right);

}


void printArray(const vector<int>& arr)
{
    for(int num : arr)
        cout << num << " ";
    cout << endl;
}


int main()
{
    vector<int> arr = {64, 45, 48, 87, 878, 78, 74, 13, 55, 44, 34, 25, 12, 22, 100, 210, 150, 124, 11, 90};

    cout << " BEFORE SORTING : ";
    printArray(arr);

    mergeSort(arr, 0, arr.size() - 1);

    cout << " AFTER SORTED : ";
    printArray(arr);

    return 0;
}
